File: tests/Scribe.Core.Tests/SilenceAutoStopTrackerTests.cs
using System.Runtime.InteropServices;
using NAudio.Wave;
using Scribe.Core.Audio;
using Xunit;

namespace Scribe.Core.Tests;

public sealed class SilenceAutoStopTrackerTests
{
    private const float Voice = 0.3f;
    private const float Quiet = 0.005f;

    [Fact]
    public void Fires_after_the_hold_window_of_silence_following_speech()
    {
        var tracker = new SilenceAutoStopTracker(startedMs: 0, silenceHoldMs: 4_000);

        Assert.False(tracker.Update(Voice, 500));    // speaking; the window measures from here
        Assert.False(tracker.Update(Quiet, 1_000));  // silence begins
        Assert.False(tracker.Update(Quiet, 4_400));  // 3.9s since the last voice; not yet
        Assert.True(tracker.Update(Quiet, 4_500));   // 4.0s; stop
    }

    [Fact]
    public void Speech_resets_the_silence_window()
    {
        var tracker = new SilenceAutoStopTracker(startedMs: 0, silenceHoldMs: 4_000);

        tracker.Update(Voice, 500);
        tracker.Update(Quiet, 3_000);
        Assert.False(tracker.Update(Voice, 4_000));  // spoke again; window restarts
        Assert.False(tracker.Update(Quiet, 7_900));
        Assert.True(tracker.Update(Quiet, 8_000));
    }

    [Fact]
    public void A_constant_level_is_steady_noise_however_loud_and_ends_on_the_lead_in()
    {
        // This used to be "continuous speech never fires", but a level that never moves is exactly
        // what steady noise looks like, and steady noise must not hold the microphone open. Real
        // speech never holds one level; the synthetic-speech test below pins that case.
        var tracker = new SilenceAutoStopTracker(startedMs: 0, silenceHoldMs: 4_000, leadInLimitMs: 10_000);

        long? stoppedAt = null;
        for (long t = 10; t <= 30_000 && stoppedAt is null; t += 10)
        {
            if (tracker.Update(Voice, t))
            {
                stoppedAt = t;
            }
        }

        Assert.Equal(10_000, stoppedAt);
        Assert.False(tracker.HeardSpeech);
    }

    [Fact]
    public void Pure_silence_fires_at_the_lead_in_limit_so_a_muted_mic_cannot_stay_hot()
    {
        var tracker = new SilenceAutoStopTracker(startedMs: 0, silenceHoldMs: 4_000, leadInLimitMs: 10_000);

        Assert.False(tracker.Update(Quiet, 9_900));
        Assert.True(tracker.Update(Quiet, 10_000));
    }

    [Fact]
    public void Reports_whether_it_ever_heard_speech_and_how_loud_the_input_got()
    {
        // The two stops look identical to the user ("it cut out after a few seconds") and need
        // opposite fixes: one is the feature working, the other is a microphone that never
        // delivered a usable level. The log says which, so this is what it reads.
        var wentQuiet = new SilenceAutoStopTracker(startedMs: 0, silenceHoldMs: 4_000);
        wentQuiet.Update(Voice, 500);
        Assert.True(wentQuiet.Update(Quiet, 4_500));
        Assert.True(wentQuiet.HeardSpeech);
        Assert.Equal(Voice, wentQuiet.PeakLevel);

        var neverHeard = new SilenceAutoStopTracker(startedMs: 0, leadInLimitMs: 10_000);
        Assert.True(neverHeard.Update(Quiet, 10_000));
        Assert.False(neverHeard.HeardSpeech);
        Assert.Equal(Quiet, neverHeard.PeakLevel);
    }

    [Fact]
    public void Peak_level_survives_a_quiet_sample_after_a_loud_one()
    {
        var tracker = new SilenceAutoStopTracker(startedMs: 0);

        tracker.Update(0.42f, 100);
        tracker.Update(Quiet, 200);

        Assert.Equal(0.42f, tracker.PeakLevel);
    }

    [Fact]
    public void Lead_in_is_longer_than_the_post_speech_hold()
    {
        // A thinking pause before the first word must not cut the dictation off at the (shorter)
        // hold window; only the lead-in limit applies until speech is heard.
        var tracker = new SilenceAutoStopTracker(startedMs: 0, silenceHoldMs: 4_000, leadInLimitMs: 10_000);

        Assert.False(tracker.Update(Quiet, 5_000));  // 5s of thinking silence; still recording
        Assert.False(tracker.Update(Voice, 6_000));  // first words arrive
        Assert.False(tracker.Update(Quiet, 9_900));
        Assert.True(tracker.Update(Quiet, 10_000));  // 4s after the last word
    }

    // Scenarios below run real 10 ms sample buffers at 48 kHz through the production meter
    // (AudioCaptureService.ComputePeak), which is what the capture service hands the tracker.
    // Every signal is synthetic and seeded, so each case is deterministic.

    [Fact]
    public void Quiet_speech_peaking_at_minus_40_dBFS_is_heard_and_stops_after_the_hold_not_the_lead_in()
    {
        // The fixed 0.02 threshold never heard this: toggle mode ended it at the 10 s lead-in while
        // the user was still talking, although VAD and the recognizer decode such audio.
        var selfNoise = Pink(seconds: 24, rmsDbfs: -75, seed: 11);
        var audio = Mix(selfNoise, SyntheticSpeech(seconds: 15, peakDbfs: -40, seed: 5), startSeconds: 1);

        var outcome = Run(audio);

        Assert.True(outcome.HeardSpeech);
        Assert.InRange(outcome.StoppedAtMs!.Value, 16_000, 20_000);
    }

    [Fact]
    public void Loud_speech_survives_a_two_second_pause_and_stops_within_the_hold_after_the_last_word()
    {
        var room = Pink(seconds: 20, rmsDbfs: -60, seed: 3);
        var audio = Mix(room, SyntheticSpeech(seconds: 5, peakDbfs: -6, seed: 8), startSeconds: 1);
        audio = Mix(audio, SyntheticSpeech(seconds: 4, peakDbfs: -6, seed: 9), startSeconds: 8);

        var outcome = Run(audio);

        Assert.True(outcome.HeardSpeech);
        Assert.InRange(outcome.StoppedAtMs!.Value, 12_000, 16_000);
    }

    [Theory]
    [InlineData(1)]
    [InlineData(2)]
    [InlineData(3)]
    public void Steady_pink_noise_at_minus_30_dBFS_is_learned_and_ends_on_the_lead_in_as_no_speech(int seed)
    {
        // The fixed threshold counted every buffer of this noise as speech and never stopped.
        var tracker = new SilenceAutoStopTracker(startedMs: 0);
        var noise = Pink(seconds: 30, rmsDbfs: -30, seed: seed);

        var outcome = Run(noise, tracker);

        Assert.Equal(10_000, outcome.StoppedAtMs);
        Assert.False(outcome.HeardSpeech);
        Assert.True(tracker.VoiceThreshold > outcome.LoudestBufferPeak);
    }

    [Fact]
    public void Speech_in_steady_noise_is_heard_and_the_noise_alone_then_lets_the_capture_end()
    {
        var noise = Pink(seconds: 25, rmsDbfs: -30, seed: 4);
        var audio = Mix(noise, SyntheticSpeech(seconds: 10, peakDbfs: -6, seed: 6), startSeconds: 2);

        var outcome = Run(audio);

        Assert.True(outcome.HeardSpeech);
        Assert.InRange(outcome.StoppedAtMs!.Value, 12_000, 16_000);
    }

    [Fact]
    public void Noise_that_starts_mid_capture_is_learned_so_it_cannot_hold_the_capture_open()
    {
        // Speech in a quiet room, then a fan starts. Until the floor climbs to the fan the noise
        // reads as voice, so the capture ends later than the speech alone would have, but it ends.
        var room = Pink(seconds: 40, rmsDbfs: -65, seed: 7);
        var audio = Mix(room, SyntheticSpeech(seconds: 3, peakDbfs: -6, seed: 10), startSeconds: 1);
        audio = Mix(audio, Pink(seconds: 34, rmsDbfs: -30, seed: 12), startSeconds: 6);

        var outcome = Run(audio);

        Assert.True(outcome.HeardSpeech);
        Assert.InRange(outcome.StoppedAtMs!.Value, 6_000, 20_000);
    }

    [Fact]
    public void Noise_that_rises_gradually_is_followed_and_never_mistaken_for_speech()
    {
        // 1 dB/s from -60 to -20 dBFS, then level. A long lead-in makes any buffer that ever read as
        // voice visible: the stop would then come 4 s after it, not exactly at the lead-in.
        var tracker = new SilenceAutoStopTracker(startedMs: 0, leadInLimitMs: 60_000);
        var noise = Ramp(Pink(seconds: 62, rmsDbfs: 0, seed: 13), fromDbfs: -60, toDbfs: -20, overSeconds: 40);

        var outcome = Run(noise, tracker);

        Assert.Equal(60_000, outcome.StoppedAtMs);
        Assert.False(outcome.HeardSpeech);
    }

    [Fact]
    public void Speech_that_fades_gradually_is_heard_well_below_the_old_threshold()
    {
        // The speaker drifts away from the microphone: -10 to -55 dBFS peak over 30 s. The old
        // fixed -34 dBFS threshold lost this about 17 s in; the absolute floor keeps it until the
        // speech drops under about -45 dBFS, roughly 24 s in.
        var room = Pink(seconds: 40, rmsDbfs: -80, seed: 14);
        var speech = Ramp(SyntheticSpeech(seconds: 30, peakDbfs: 0, seed: 15), fromDbfs: -10, toDbfs: -55, overSeconds: 30);
        var audio = Mix(room, speech, startSeconds: 1);

        var outcome = Run(audio);

        Assert.True(outcome.HeardSpeech);
        Assert.InRange(outcome.StoppedAtMs!.Value, 24_000, 31_000);
    }

    [Fact]
    public void Digital_silence_ends_on_the_lead_in_as_no_speech()
    {
        var outcome = Run(new float[15 * SampleRate]);

        Assert.Equal(10_000, outcome.StoppedAtMs);
        Assert.False(outcome.HeardSpeech);
        Assert.Equal(0f, outcome.LoudestBufferPeak);
    }

    [Fact]
    public void Synthetic_speech_with_natural_gaps_never_stops_however_long_it_runs()
    {
        var room = Pink(seconds: 60, rmsDbfs: -60, seed: 16);
        var audio = Mix(room, SyntheticSpeech(seconds: 60, peakDbfs: -12, seed: 17), startSeconds: 0);

        var outcome = Run(audio);

        Assert.Null(outcome.StoppedAtMs);
        Assert.True(outcome.HeardSpeech);
    }

    // In the collection that runs alone (stream TR, item 1): nothing else in the process runs while it measures.
    [Collection(AllocationMeasurementCollection.Name)]
    public sealed class Allocations
    {
        [Fact]
        public void Updating_allocates_nothing()
        {
            // It runs once per capture buffer on the audio thread, for the whole dictation.
            var tracker = new SilenceAutoStopTracker(startedMs: 0, leadInLimitMs: long.MaxValue);
            for (var i = 1; i <= 1_000; i++)
            {
                tracker.Update(0.1f * (i % 7), i * 10L);
            }

            _ = RuntimeWork.Now().Since(RuntimeWork.Now());

            // The tracker allocates nothing, but with the whole suite running in parallel the runtime has been
            // seen to allocate about 3 KB on this thread during one measured pass (3 full-suite runs in 20,
            // never with this class alone). That is one-time work, while an allocation in Update recurs on
            // every pass of 100,000 buffers, so one clean pass proves the claim and a real allocation still
            // fails all five. (Stream TR kept these passes as they were, and runs the test alone.)
            var allocated = long.MaxValue;
            var during = default(RuntimeWork);
            var next = 1_001;
            for (var pass = 0; pass < 5 && allocated != 0; pass++)
            {
                var work = RuntimeWork.Now();
                var before = GC.GetAllocatedBytesForCurrentThread();
                Pass(next);
                allocated = GC.GetAllocatedBytesForCurrentThread() - before;
                during = RuntimeWork.Now().Since(work);
                next += 100_000;
            }

            AllocationMeasurement.AssertZero(allocated, during, "The last of five passes of 100,000 updates", () => Pass(next));

            void Pass(int first)
            {
                for (var i = first; i < first + 100_000; i++)
                {
                    tracker.Update(i % 3 == 0 ? 0f : 0.05f * (i % 11), i * 10L);
                }
            }
        }
    }

    [Theory]
    [InlineData(0f)]
    [InlineData(-0.1f)]
    [InlineData(1.5f)]
    [InlineData(float.NaN)]
    public void An_absolute_floor_outside_the_level_range_is_rejected(float absoluteFloor)
    {
        Assert.Throws<ArgumentOutOfRangeException>(() => new SilenceAutoStopTracker(0, absoluteFloor));
    }

    private const int SampleRate = 48_000;
    private const int SamplesPerBuffer = SampleRate / 100;
    private static readonly WaveFormat MonoFloat48k = WaveFormat.CreateIeeeFloatWaveFormat(SampleRate, 1);

    private sealed record Outcome(long? StoppedAtMs, bool HeardSpeech, float LoudestBufferPeak);

    private static Outcome Run(float[] samples, SilenceAutoStopTracker? tracker = null)
    {
        tracker ??= new SilenceAutoStopTracker(startedMs: 0);
        var loudest = 0f;
        for (var buffer = 0; (buffer + 1) * SamplesPerBuffer <= samples.Length; buffer++)
        {
            var span = samples.AsSpan(buffer * SamplesPerBuffer, SamplesPerBuffer);
            var peak = AudioCaptureService.ComputePeak(MemoryMarshal.AsBytes(span), MonoFloat48k);
            loudest = Math.Max(loudest, peak);
            var now = (buffer + 1) * 10L;
            if (tracker.Update(peak, now))
            {
                return new Outcome(now, tracker.HeardSpeech, loudest);
            }
        }

        return new Outcome(null, tracker.HeardSpeech, loudest);
    }

    /// <summary>Pink noise (Paul Kellet's filter over seeded Gaussian white noise) at an RMS level.</summary>
    private static float[] Pink(int seconds, double rmsDbfs, int seed)
    {
        var random = new Random(seed);
        double b0 = 0, b1 = 0, b2 = 0, b3 = 0, b4 = 0, b5 = 0, b6 = 0;
        var samples = new float[seconds * SampleRate];
        for (var i = 0; i < samples.Length; i++)
        {
            var white = Math.Sqrt(-2 * Math.Log(1 - random.NextDouble())) * Math.Cos(2 * Math.PI * random.NextDouble());
            b0 = (0.99886 * b0) + (white * 0.0555179);
            b1 = (0.99332 * b1) + (white * 0.0750759);
            b2 = (0.96900 * b2) + (white * 0.1538520);
            b3 = (0.86650 * b3) + (white * 0.3104856);
            b4 = (0.55000 * b4) + (white * 0.5329522);
            b5 = (-0.7616 * b5) - (white * 0.0168980);
            samples[i] = (float)(b0 + b1 + b2 + b3 + b4 + b5 + b6 + (white * 0.5362));
            b6 = white * 0.115926;
        }

        var rms = Math.Sqrt(samples.Average(sample => (double)sample * sample));
        var gain = (float)(Math.Pow(10, rmsDbfs / 20) / rms);
        for (var i = 0; i < samples.Length; i++)
        {
            samples[i] *= gain;
        }

        return samples;
    }

    /// <summary>
    /// Speech-shaped signal: voiced syllables of 120 to 260 ms on a 110 to 180 Hz harmonic
    /// complex, each rising and decaying, with a spread of 12 dB between syllables, short gaps
    /// inside words and longer ones between them, scaled so its loudest sample sits at the peak.
    /// </summary>
    private static float[] SyntheticSpeech(int seconds, double peakDbfs, int seed)
    {
        var random = new Random(seed);
        var samples = new float[seconds * SampleRate];
        var position = 0;
        var syllable = 0;
        while (position < samples.Length)
        {
            var length = (int)(SampleRate * (0.12 + (random.NextDouble() * 0.14)));
            var amplitude = Math.Pow(10, (random.NextDouble() - 0.5) * 12 / 20);
            var pitch = 110 + (random.NextDouble() * 70);
            for (var i = 0; i < length && position + i < samples.Length; i++)
            {
                var phase = 2 * Math.PI * pitch * (position + i) / SampleRate;
                var voice = Math.Sin(phase) + (0.5 * Math.Sin(2 * phase)) + (0.25 * Math.Sin(3 * phase));
                samples[position + i] = (float)(amplitude * Math.Sin(Math.PI * i / length) * voice);
            }

            position += length;
            var wordEnds = ++syllable % 3 == 0;
            position += (int)(SampleRate * (wordEnds ? 0.15 + (random.NextDouble() * 0.2) : 0.03 + (random.NextDouble() * 0.06)));
        }

        var loudest = samples.Max(Math.Abs);
        var gain = (float)(Math.Pow(10, peakDbfs / 20) / loudest);
        for (var i = 0; i < samples.Length; i++)
        {
            samples[i] *= gain;
        }

        return samples;
    }

    /// <summary>Applies a gain that moves linearly in dB from one level to another, then holds.</summary>
    private static float[] Ramp(float[] samples, double fromDbfs, double toDbfs, double overSeconds)
    {
        var ramped = new float[samples.Length];
        for (var i = 0; i < samples.Length; i++)
        {
            var progress = Math.Min(1, i / (overSeconds * SampleRate));
            ramped[i] = (float)(samples[i] * Math.Pow(10, (fromDbfs + ((toDbfs - fromDbfs) * progress)) / 20));
        }

        return ramped;
    }

    private static float[] Mix(float[] bed, float[] overlay, double startSeconds)
    {
        var mixed = (float[])bed.Clone();
        var offset = (int)(startSeconds * SampleRate);
        for (var i = 0; i < overlay.Length && offset + i < mixed.Length; i++)
        {
            mixed[offset + i] += overlay[i];
        }

        return mixed;
    }
}

File: macos/Scribe/Tests/ScribeTests/SilenceAutoStopTrackerTests.swift
import XCTest

@testable import Scribe

/// A port of Windows' `SilenceAutoStopTrackerTests`. The scenarios run real 10 ms blocks of 48 kHz audio
/// through the tracker, the block size `CaptureProcessor` feeds it at that rate. Every signal is synthetic and
/// seeded, so each case is deterministic.
final class SilenceAutoStopTrackerTests: XCTestCase {
    private let voice: Float = 0.3
    private let quiet: Float = 0.005

    private func makeTracker(
        silenceHold: Duration = .seconds(4),
        leadIn: Duration = .seconds(10)
    ) throws -> SilenceAutoStopTracker {
        let configuration = try XCTUnwrap(SilenceAutoStopConfiguration(silenceHold: silenceHold, leadInLimit: leadIn))
        return SilenceAutoStopTracker(configuration: configuration)
    }

    func testFiresAfterTheHoldWindowOfSilenceFollowingSpeech() throws {
        var tracker = try makeTracker()

        XCTAssertFalse(tracker.update(level: voice, atMilliseconds: 500))
        XCTAssertFalse(tracker.update(level: quiet, atMilliseconds: 1_000))
        XCTAssertFalse(tracker.update(level: quiet, atMilliseconds: 4_400))
        XCTAssertTrue(tracker.update(level: quiet, atMilliseconds: 4_500))
    }

    func testSpeechResetsTheSilenceWindow() throws {
        var tracker = try makeTracker()

        _ = tracker.update(level: voice, atMilliseconds: 500)
        _ = tracker.update(level: quiet, atMilliseconds: 3_000)
        XCTAssertFalse(tracker.update(level: voice, atMilliseconds: 4_000))
        XCTAssertFalse(tracker.update(level: quiet, atMilliseconds: 7_900))
        XCTAssertTrue(tracker.update(level: quiet, atMilliseconds: 8_000))
    }

    /// A level that never moves is what steady noise looks like, however loud, and steady noise must not hold
    /// the microphone open. Real speech never holds one level; the synthetic speech test pins that case.
    func testAConstantLevelIsSteadyNoiseHoweverLoudAndEndsOnTheLeadIn() throws {
        var tracker = try makeTracker()

        var stoppedAt: Int64?
        var timestamp: Int64 = 10
        while timestamp <= 30_000, stoppedAt == nil {
            if tracker.update(level: voice, atMilliseconds: timestamp) {
                stoppedAt = timestamp
            }
            timestamp += 10
        }

        XCTAssertEqual(stoppedAt, 10_000)
        XCTAssertFalse(tracker.heardSpeech)
    }

    func testPureSilenceFiresAtTheLeadInLimitSoAMutedMicrophoneCannotStayOn() throws {
        var tracker = try makeTracker()

        XCTAssertFalse(tracker.update(level: quiet, atMilliseconds: 9_900))
        XCTAssertTrue(tracker.update(level: quiet, atMilliseconds: 10_000))
    }

    func testReportsWhetherItEverHeardSpeechAndHowLoudTheInputGot() throws {
        var wentQuiet = try makeTracker()
        _ = wentQuiet.update(level: voice, atMilliseconds: 500)
        XCTAssertTrue(wentQuiet.update(level: quiet, atMilliseconds: 4_500))
        XCTAssertTrue(wentQuiet.heardSpeech)
        XCTAssertEqual(wentQuiet.peakLevel, voice)

        var neverHeard = try makeTracker()
        XCTAssertTrue(neverHeard.update(level: quiet, atMilliseconds: 10_000))
        XCTAssertFalse(neverHeard.heardSpeech)
        XCTAssertEqual(neverHeard.peakLevel, quiet)
    }

    func testPeakLevelSurvivesAQuietBlockAfterALoudOne() throws {
        var tracker = try makeTracker()

        _ = tracker.update(level: 0.42, atMilliseconds: 100)
        _ = tracker.update(level: quiet, atMilliseconds: 200)

        XCTAssertEqual(tracker.peakLevel, 0.42)
    }

    /// A thinking pause before the first word must not end the recording at the shorter hold window.
    func testTheLeadInIsLongerThanThePostSpeechHold() throws {
        var tracker = try makeTracker()

        XCTAssertFalse(tracker.update(level: quiet, atMilliseconds: 5_000))
        XCTAssertFalse(tracker.update(level: voice, atMilliseconds: 6_000))
        XCTAssertFalse(tracker.update(level: quiet, atMilliseconds: 9_900))
        XCTAssertTrue(tracker.update(level: quiet, atMilliseconds: 10_000))
    }

    func testAnAbsoluteFloorOutsideTheLevelRangeOrANegativeDurationIsRejected() {
        for floor: Float in [0, -0.1, 1.5, .nan] {
            XCTAssertNil(SilenceAutoStopConfiguration(absoluteFloor: floor), "\(floor)")
        }
        XCTAssertNil(SilenceAutoStopConfiguration(silenceHold: .seconds(-1)))
        XCTAssertNil(SilenceAutoStopConfiguration(leadInLimit: .seconds(-1)))
        XCTAssertNotNil(SilenceAutoStopConfiguration(absoluteFloor: 1))
    }

    func testAnEndlessLeadInNeverFiresWithoutSpeech() throws {
        var tracker = try makeTracker(leadIn: .seconds(Int64.max))

        XCTAssertFalse(tracker.update(level: quiet, atMilliseconds: 86_400_000))
    }

    // MARK: - Scenarios on real 10 ms blocks of 48 kHz audio

    func testQuietSpeechPeakingAtMinus40DbfsIsHeardAndStopsAfterTheHoldNotTheLeadIn() {
        let selfNoise = AudioTestSignal.pink(seconds: 24, rmsDbfs: -75, seed: 11)
        let audio = AudioTestSignal.mix(
            selfNoise, AudioTestSignal.syntheticSpeech(seconds: 15, peakDbfs: -40, seed: 5), startSeconds: 1)

        let outcome = run(audio)

        XCTAssertTrue(outcome.heardSpeech)
        XCTAssertNotNil(outcome.stoppedAt)
        assertStopped(outcome, within: 16_000...20_000)
    }

    func testLoudSpeechSurvivesATwoSecondPauseAndStopsWithinTheHoldAfterTheLastWord() {
        let room = AudioTestSignal.pink(seconds: 20, rmsDbfs: -60, seed: 3)
        var audio = AudioTestSignal.mix(
            room, AudioTestSignal.syntheticSpeech(seconds: 5, peakDbfs: -6, seed: 8), startSeconds: 1)
        audio = AudioTestSignal.mix(
            audio, AudioTestSignal.syntheticSpeech(seconds: 4, peakDbfs: -6, seed: 9), startSeconds: 8)

        let outcome = run(audio)

        XCTAssertTrue(outcome.heardSpeech)
        assertStopped(outcome, within: 12_000...16_000)
    }

    /// A fixed threshold counted every block of this noise as speech and never stopped.
    func testSteadyPinkNoiseAtMinus30DbfsIsLearnedAndEndsOnTheLeadInAsNoSpeech() {
        for seed: UInt64 in [1, 2, 3] {
            var tracker = SilenceAutoStopTracker()
            let noise = AudioTestSignal.pink(seconds: 30, rmsDbfs: -30, seed: seed)

            let outcome = run(noise, tracker: &tracker)

            XCTAssertEqual(outcome.stoppedAt, 10_000, "seed \(seed)")
            XCTAssertFalse(outcome.heardSpeech, "seed \(seed)")
            XCTAssertGreaterThan(tracker.voiceThreshold, outcome.loudestBlockPeak, "seed \(seed)")
        }
    }

    func testSpeechInSteadyNoiseIsHeardAndTheNoiseAloneThenLetsTheRecordingEnd() {
        let noise = AudioTestSignal.pink(seconds: 25, rmsDbfs: -30, seed: 4)
        let audio = AudioTestSignal.mix(
            noise, AudioTestSignal.syntheticSpeech(seconds: 10, peakDbfs: -6, seed: 6), startSeconds: 2)

        let outcome = run(audio)

        XCTAssertTrue(outcome.heardSpeech)
        assertStopped(outcome, within: 12_000...16_000)
    }

    /// Speech in a quiet room, then a fan starts. Until the floor climbs to the fan the noise reads as voice, so
    /// the recording ends later than the speech alone would have, but it ends.
    func testNoiseThatStartsMidRecordingIsLearnedSoItCannotHoldTheRecordingOpen() {
        let room = AudioTestSignal.pink(seconds: 40, rmsDbfs: -65, seed: 7)
        var audio = AudioTestSignal.mix(
            room, AudioTestSignal.syntheticSpeech(seconds: 3, peakDbfs: -6, seed: 10), startSeconds: 1)
        audio = AudioTestSignal.mix(audio, AudioTestSignal.pink(seconds: 34, rmsDbfs: -30, seed: 12), startSeconds: 6)

        let outcome = run(audio)

        XCTAssertTrue(outcome.heardSpeech)
        assertStopped(outcome, within: 6_000...20_000)
    }

    /// 1 dB/s from -60 to -20 dBFS, then level. With a long lead-in, any block that ever read as voice would show:
    /// the stop would then come 4 s after it, not exactly at the lead-in.
    func testNoiseThatRisesGraduallyIsFollowedAndNeverMistakenForSpeech() throws {
        var tracker = try makeTracker(leadIn: .seconds(60))
        let noise = AudioTestSignal.ramp(
            AudioTestSignal.pink(seconds: 62, rmsDbfs: 0, seed: 13), fromDbfs: -60, toDbfs: -20, overSeconds: 40)

        let outcome = run(noise, tracker: &tracker)

        XCTAssertEqual(outcome.stoppedAt, 60_000)
        XCTAssertFalse(outcome.heardSpeech)
    }

    /// The speaker drifts away from the microphone: -10 to -55 dBFS peak over 30 s. The absolute floor keeps the
    /// speech until it drops under about -45 dBFS, roughly 24 s in.
    func testSpeechThatFadesGraduallyIsHeardWellBelowTheOldThreshold() {
        let room = AudioTestSignal.pink(seconds: 40, rmsDbfs: -80, seed: 14)
        let speech = AudioTestSignal.ramp(
            AudioTestSignal.syntheticSpeech(seconds: 30, peakDbfs: 0, seed: 15), fromDbfs: -10, toDbfs: -55,
            overSeconds: 30)
        let audio = AudioTestSignal.mix(room, speech, startSeconds: 1)

        let outcome = run(audio)

        XCTAssertTrue(outcome.heardSpeech)
        assertStopped(outcome, within: 24_000...31_000)
    }

    func testDigitalSilenceEndsOnTheLeadInAsNoSpeech() {
        let outcome = run([Float](repeating: 0, count: 15 * AudioTestSignal.sampleRate))

        XCTAssertEqual(outcome.stoppedAt, 10_000)
        XCTAssertFalse(outcome.heardSpeech)
        XCTAssertEqual(outcome.loudestBlockPeak, 0)
    }

    func testSyntheticSpeechWithNaturalGapsNeverStopsHoweverLongItRuns() {
        let room = AudioTestSignal.pink(seconds: 60, rmsDbfs: -60, seed: 16)
        let audio = AudioTestSignal.mix(
            room, AudioTestSignal.syntheticSpeech(seconds: 60, peakDbfs: -12, seed: 17), startSeconds: 0)

        let outcome = run(audio)

        XCTAssertNil(outcome.stoppedAt)
        XCTAssertTrue(outcome.heardSpeech)
    }

    // MARK: - Helpers

    private func assertStopped(
        _ outcome: Outcome, within range: ClosedRange<Int64>, file: StaticString = #filePath, line: UInt = #line
    ) {
        guard let stoppedAt = outcome.stoppedAt else {
            return XCTFail("never stopped", file: file, line: line)
        }
        XCTAssertTrue(range.contains(stoppedAt), "stopped at \(stoppedAt)", file: file, line: line)
    }

    private struct Outcome {
        let stoppedAt: Int64?
        let heardSpeech: Bool
        let loudestBlockPeak: Float
    }

    private func run(_ samples: [Float]) -> Outcome {
        var tracker = SilenceAutoStopTracker()
        return run(samples, tracker: &tracker)
    }

    private func run(_ samples: [Float], tracker: inout SilenceAutoStopTracker) -> Outcome {
        let blockLength = AudioTestSignal.sampleRate / 100
        var loudest: Float = 0
        var block = 0
        while (block + 1) * blockLength <= samples.count {
            var peak: Float = 0
            for index in (block * blockLength)..<((block + 1) * blockLength) {
                peak = max(peak, abs(samples[index]))
            }
            loudest = max(loudest, peak)
            let timestamp = Int64(block + 1) * 10
            if tracker.update(level: peak, atMilliseconds: timestamp) {
                return Outcome(stoppedAt: timestamp, heardSpeech: tracker.heardSpeech, loudestBlockPeak: loudest)
            }
            block += 1
        }
        return Outcome(stoppedAt: nil, heardSpeech: tracker.heardSpeech, loudestBlockPeak: loudest)
    }
}

final class CaptureStopPolicyTests: XCTestCase {
    func testAHoldNeverStopsOnSilenceAndBothStopAtTheTenMinuteCeiling() {
        XCTAssertNil(CaptureStopPolicy.hold().silence)
        XCTAssertEqual(CaptureStopPolicy.toggle().silence, .standard)
        XCTAssertEqual(CaptureStopPolicy.hold().maximumDuration, .seconds(600))
        XCTAssertEqual(CaptureStopPolicy.toggle().maximumDuration, .seconds(600))
    }

    func testThePolicyFollowsTheGestureOfTheBindingThatFired() {
        XCTAssertEqual(CaptureStopPolicy(gesture: .hold), .hold())
        XCTAssertEqual(CaptureStopPolicy(gesture: .toggle), .toggle())
        XCTAssertFalse(CaptureStopPolicy(gesture: .toggle, autoStopOnSilence: false).stopsOnSilence)
        XCTAssertEqual(HotkeyBinding(keyCode: HotkeySettingsStore.capsLockKeyCode).gesture, .toggle)
    }

    func testAZeroOrNegativeCeilingMeansNoCeiling() {
        XCTAssertNil(CaptureStopPolicy.hold(maximumDuration: .zero).maximumDuration)
        XCTAssertNil(CaptureStopPolicy.toggle(maximumDuration: .seconds(-5)).maximumDuration)
        XCTAssertNil(CaptureStopPolicy.hold(maximumDuration: nil).maximumDuration)
    }

    func testTheStandardConfigurationIsWindowsDefaults() {
        let standard = SilenceAutoStopConfiguration.standard
        XCTAssertEqual(standard.absoluteFloor, 0.0056)
        XCTAssertEqual(standard.silenceHold, .seconds(4))
        XCTAssertEqual(standard.leadInLimit, .seconds(10))
    }
}
